pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/BigIntegerInstantiationRule.java

```diff
@@ -6,57 +6,65 @@
+import net.sourceforge.pmd.util.CollectionUtil;
 
 /**
- * Rule that marks instantiations of new {@link BigInteger} or
- * {@link BigDecimal} objects, when there is a well-known constant available,
- * such as {@link BigInteger#ZERO}.
+ * Rule that marks instantiations of new {@link BigInteger} or {@link BigDecimal} objects, when there is a well-known
+ * constant available, such as {@link BigInteger#ZERO}.
  */
-public class BigIntegerInstantiationRule extends AbstractJavaRule {
+public class BigIntegerInstantiationRule extends AbstractJavaRulechainRule {
 
-    @Override
-    public Object visit(ASTAllocationExpression node, Object data) {
-        Node type = node.getChild(0);
 
-        if (!(type instanceof ASTClassOrInterfaceType)) {
-            return super.visit(node, data);
-        }
+    private static final Set<String> CONSTANTS = CollectionUtil.setOf("0", "0.", "1");
 
-        boolean jdk15 = node.getAstInfo().getLanguageVersion().compareTo(LanguageRegistry.getLanguage(JavaLanguageModule.NAME).getVersion("1.5")) >= 0;
-        if ((TypeTestUtil.isA(BigInteger.class, (ASTClassOrInterfaceType) type)
-                || jdk15 && TypeTestUtil.isA(BigDecimal.class, (ASTClassOrInterfaceType) type))
-                && !node.hasDescendantOfType(ASTArrayDimsAndInits.class)) {
-            ASTArguments args = node.getFirstChildOfType(ASTArguments.class);
-            if (args.size() == 1) {
-                ASTLiteral literal = node.getFirstDescendantOfType(ASTLiteral.class);
-                if (literal == null
-                        || literal.getParent().getParent().getParent().getParent().getParent() != args) {
-                    return super.visit(node, data);
-                }
+    public BigIntegerInstantiationRule() {
+        super(ASTConstructorCall.class);
+    }
 
-                String img = literal.getImage();
-                if (literal.isStringLiteral()) {
-                    img = img.substring(1, img.length() - 1);
-                }
 
-                if ("0".equals(img) || "1".equals(img) || jdk15 && "10".equals(img)) {
-                    addViolation(data, node);
-                    return data;
+    @Override
+    public Object visit(ASTConstructorCall node, Object data) {
+        LanguageVersion languageVersion = node.getAstInfo().getLanguageVersion();
+        boolean jdk15 = languageVersion
+                .compareTo(LanguageRegistry.getLanguage(JavaLanguageModule.NAME).getVersion("1.5")) >= 0;
+        boolean jdk9 = languageVersion
+                .compareTo(LanguageRegistry.getLanguage(JavaLanguageModule.NAME).getVersion("9")) >= 0;
+
+        if (TypeTestUtil.isA(BigInteger.class, node) || jdk15 && TypeTestUtil.isA(BigDecimal.class, node)) {
+
+            @NonNull
+            ASTArgumentList arguments = node.getArguments();
+            if (arguments.size() == 1) {
+                ASTExpression firstArg = arguments.get(0);
+                if (firstArg instanceof ASTStringLiteral) {
```

Comment:
I think we should probably just check `getConstValue`, and avoid testing the type of the argument node. This would be slightly more general, and the rule would be simpler. It would eg catch 
```
class Foo {
    static final String Z = "0";
    static { new BigInteger(Z); }
}

Reply:
yes, that works. Thanks for the suggestion!